---
title: Applied Statistical Methods - Notebook 7
author: Peter von Rohr
date: '2023-05-14'
output: html_notebook
params:
  doctype:
    label: Document Type
    value: solution
    choices:
    - exercise
    - solution
    - notebook
  isonline:
    label: Online (y/n)
    value: true
    choices:
    - true
    - false
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Problem 1: Marker Effects Model
```{r ex11p01-setup, echo=FALSE}
s_ex11_p01_data_path <- "https://charlotte-ngs.github.io/asmss2023/data/asm_geno_sim_data.csv"

sigma_q2 <- 3
sigma_e2 <- 36
```

Predict genomic breeding values using a marker effects model. The dataset is available from

```{r, echo=FALSE}
cat(s_ex11_p01_data_path, "\n")
```

### Hints

* The variance $\sigma_q^2$ of the marker effect is $`r sigma_q2`$.
* The residual variance $\sigma_e^2$ is $`r sigma_e2`$
* The sex of each animal can be modelled as a fixed effect

### Your Solution

* Read the data
* Setup mixed model equations to predict marker effects for all the SNP-loci
* Compute predicted genomic breeding values based on the estimated marker effects





## Problem 2: Breeding Value Based Model
```{r ex11p02-setup, echo=FALSE}
s_ex11_p02_data_path <- "https://charlotte-ngs.github.io/asmss2023/data/asm_geno_sim_data.csv"

sigma_q2 <- 3
sigma_u2 <- 3*sigma_q2
sigma_e2 <- 36
```

Use the same dataset as in Problem 1 to predict genomic breeding values based on a breeding-value model. The dataset is available from

```{r, echo=FALSE}
cat(s_ex11_p02_data_path, "\n")
```

### Hints

* The genomic variance $\sigma_u^2$ of the marker effect is $`r sigma_u2`$.
* The residual variance $\sigma_e^2$ is $`r sigma_e2`$
* The sex of each animal can be modelled as a fixed effect
* Use the following function to compute the genomic relationship matrix $G$ based on the matrix of genotypes

```{r}
computeMatGrm <- function(pmatData) {
  matData <- pmatData
  # check the coding, if matData is -1, 0, 1 coded, then add 1 to get to 0, 1, 2 coding
  if (min(matData) < 0) matData <- matData + 1
  # Allele frequencies, column vector of P and sum of frequency products
  freq <- apply(matData, 2, mean) / 2
  P <- 2 * (freq - 0.5)
  sumpq <- sum(freq*(1-freq))
  # Changing the coding from (0,1,2) to (-1,0,1) and subtract matrix P
  Z <- matData - 1 - matrix(P, nrow = nrow(matData), 
                             ncol = ncol(matData), 
                             byrow = TRUE)
  # Z%*%Zt is replaced by tcrossprod(Z)
  return(tcrossprod(Z)/(2*sumpq))
}
```

* If the genomic relationship matrix $G$ which is computed by the function above cannot be inverted, add $0.05 * I$ to $G$ which results in $G^*$ and use $G^*$ as genomic relationship matrix.


### Your Solution

* Read the data
* Compute the inverse genomic relationship matrix using the given function for the genomic relationship matrix
* Setup mixed model equations to predict genomic breeding values






```{r, echo=FALSE, results='asis'}
cat('\n---\n\n _Latest Changes: ', format(Sys.time(), '%Y-%m-%d %H:%M:%S'), ' (', Sys.info()['user'], ')_\n', sep = '')
```
 
